core/state/state_object.go

```diff
@@ -202,6 +205,13 @@ func (self *stateObject) updateTrie(db Database) Trie {
 	tr := self.getTrie(db)
 	for key, value := range self.dirtyStorage {
 		delete(self.dirtyStorage, key)
+
+		// Skip noop changes, persist actual changes
+		if value == self.originStorage[key] {
```

Comment:
If user deletes some variables and makes the value empty, the `delete` operation could be ignored here.